### PR TITLE
Allow to create or update objects with RealmOptional as primary key

### LINKED ISSUE #3671: RealmSwift.Object with numeric optional primary keys can't be added with update=true
## Goals

Add an object of a class defined from Swift with a numeric optional primary key to a Realm.

## Expected Results

Successfully creating the object.

## Actual Results

>** Terminating app due to uncaught exception 'NSUnknownKeyException', reason: '[ valueForUndefinedKey:]: this class is not key value coding-compliant for the key (null).'

## Code Sample

```swift
class Post: Object {
    let postId = RealmOptional<Int>()
    dynamic var postImage : PostImage?
    dynamic var postUser: User?

    override class func primaryKey() -> String? {
        return "postId"
    }
 }

class PostTest : XCTestCase {
   func testExample() {
        let realm = try! Realm()
        realm.beginWrite()
        let post = Post()
        post.postId.value = json["image"]["image_id"].intValue
        realm.add(post, update: true)
        try! realm.commitWrite()
   }
}

```

### PR DESCRIPTION
**Pull request by mrackwitz**
Fixes #3671.

/c @kishikawakatsumi @jpsim

### CODE REVIEW

**Comment by kishikawakatsumi** [comment]
👍

File: Realm/RLMObjectStore.mm [hunk 1]
```diff
@@ -182,7 +182,12 @@ void RLMAddObjectToRealm(__unsafe_unretained RLMObjectBase *const object,
 
     // get or create row
     bool created;
-    auto primaryGetter = [=](__unsafe_unretained RLMProperty *const p) { return [object valueForKey:p.getterName]; };
+    auto primaryGetter = [=](__unsafe_unretained RLMProperty *const prop) {
+        if (prop.swiftIvar) {
+            return static_cast<RLMOptionalBase *>(object_getIvar(object, prop.swiftIvar)).underlyingValue;
+        }
+        return [object valueForKey:prop.getterName];
```
Comment (kishikawakatsumi) [inline Realm/RLMObjectStore.mm]: It seems that `[object valueForKey:prop.getterName]` should be `[object valueForKey:prop.name]`?

Comment (mrackwitz) [inline Realm/RLMObjectStore.mm]: Yep, that's better. :+1:

File: Realm/RLMObjectStore.mm [hunk 2]
```diff
@@ -182,7 +182,7 @@ void RLMAddObjectToRealm(__unsafe_unretained RLMObjectBase *const object,
 
     // get or create row
     bool created;
-    auto primaryGetter = [=](__unsafe_unretained RLMProperty *const p) { return [object valueForKey:p.getterName]; };
+    auto primaryGetter = [=](__unsafe_unretained RLMProperty *const p) { return  [object valueForKey:p.name]; };
```
Comment (kishikawakatsumi) [inline Realm/RLMObjectStore.mm]: Extra whitespace?

**Comment by jpsim** [comment]
👍

**Pull request merged**